source/code/modules/iceshard_pipelines/private/pipeline_ui/ip_ui_oven_elements.cxx

```diff
@@ -0,0 +1,398 @@
+            if (result)
+            {
+                inout_str = inout_str.substr(type_end + 1);
+
+                ice::usize const action_value_end = inout_str.find_first_of(u8",} ");
+                out_action.action_value = inout_str.substr(0, action_value_end);
+                result = out_action.action_value.empty() == false;
+
+                // Move to the next expected token.
+                inout_str = inout_str.substr(action_value_end + 1);
+            }
+        }
+        return result;
+    }
+
+    bool parse_data_reference(
+        ice::Utf8String str,
+        ice::RawData& out_action
+    ) noexcept
+    {
+        out_action.data_type = ice::ui::DataSource::None;
+        out_action.data_source = {};
+
+        bool result = false;
+        if (remove_brackets(str))
+        {
+            ice::usize const type_end = str.find_first_of(u8' ');
+            ice::Utf8String const data_type = str.substr(0, type_end);
+
+            if (data_type == RawAction::Constant_ActionDataType_Resource)
+            {
+                out_action.data_type = ice::ui::DataSource::ValueResource;
+                out_action.data_source = str.substr(type_end + 1);
+            }
+            else if (data_type == RawAction::Constant_ActionDataType_Property)
+            {
+                out_action.data_type = ice::ui::DataSource::ValueProperty;
+                out_action.data_source = str.substr(type_end + 1);
+            }
+            else if (data_type == RawAction::Constant_ActionDataType_UIPage)
+            {
+                out_action.data_type = ice::ui::DataSource::ValueUIPage;
+                out_action.data_source = str.substr(type_end + 1);
+            }
+
+            result = out_action.data_type != ice::ui::DataSource::None;
+            if (result)
+            {
+                ice::trim(out_action.data_source, ' ');
+            }
+        }
+
+        return result;
+    }
+
+    bool parse_action_data(
+        ice::Utf8String& inout_str,
+        ice::RawAction& out_action
+    ) noexcept
+    {
+        out_action.data.data_type = ice::ui::DataSource::None;
+        out_action.data.data_source = {};
+
+        bool result = true;
+        ice::usize const data_start = inout_str.find_first_of('=');
+        if (data_start != ice::String::npos)
+        {
+            ice::usize const data_arg_start = inout_str.find_last_of(u8" ,", data_start);
+            ice::Utf8String const data_arg = inout_str.substr(data_arg_start + 1, (data_start - data_arg_start) - 1);
+
+            inout_str = inout_str.substr(data_start + 1);
+            if (remove_brackets(inout_str))
+            {
+                ICE_ASSERT(false, "TODO!");
+                //ice::usize const type_end = str.find_first_of(u8' ');
+                //ice::Utf8String const data_type = str.substr(0, type_end);
+
+                //if (out_action.action_type == ice::ui::ActionType::Shard
```

Comment:
This might be a follow up issues, as currently there is no use case to have this solved.

Reply:
For now there is no known situation where we would need to send additional data with shards.
This issues might need a revisit at a later point. #134